drivers/bluetooth/iton_bt.c

```diff
@@ -0,0 +1,245 @@
+        case batt_voltage_low:
+            iton_bt_battery_voltage_low();
+            break;
+        case batt_exit_low_battery_mode:
+            iton_bt_battery_exit_low_battery_mode();
+            break;
+        case batt_low_power_shutdown:
+            iton_bt_battery_low_power_shutdown();
+            break;
+        case batt_above_70:
+        case batt_between_30_70:
+        case batt_below_30:
+            iton_bt_battery_level(data);
+            break;
+        case batt_wake_mcu:
+            #ifdef ITON_BT_ENABLE_ACK
+            iton_bt_send_ack(control_bt, wake_ack);
+            #endif
+            break;
+        case batt_unknown:
+            #ifdef ITON_BT_ENABLE_ACK
+            iton_bt_send_ack(control_bt, unknown_ack);
+            #endif
+            break;
+        case query_working_mode:
+            break;
+        case query_bt_name:
+            break;
+    }
+}
+
+static inline void iton_bt_rx_bluetooth_notif(uint8_t data) {
+    switch (iton_bt_buffer[2]) {
+        case bt_connection_success:
+            iton_bt_is_connected = true;
+
+            #ifdef ITON_BT_ENABLE_ACK
+            iton_bt_send_ack(control_bt, connect_ack);
+            #endif
+
+            iton_bt_connection_successful();
+            break;
+        case bt_entered_pairing:
+            iton_bt_entered_pairing();
+            break;
+        case bt_disconected:
+            iton_bt_is_connected = false;
+
+            #ifdef ITON_BT_ENABLE_ACK
+            iton_bt_send_ack(control_bt, disconnect_ack);
+            #endif
+
+            iton_bt_disconnected();
+            break;
+        case bt_enters_connection:
+            iton_bt_enters_connection_state();
+            break;
+    }
+}
+
+static void iton_bt_rx_cb(void *arg) {
+    if (readPin(ITON_BT_INT_LINE)) {
+        chSysLockFromISR();
+        spiStartReceiveI(&ITON_BT_SPI_PORT, ITON_BT_BUFFER_LEN, &iton_bt_buffer[0]);
+        chSysUnlockFromISR();
+    } else {
+        chSysLockFromISR();
+        spiStopTransferI(&ITON_BT_SPI_PORT, NULL);
+        chSysUnlockFromISR();
+
+        #ifdef ITON_BT_ENABLE_ACK
+        // hack to make sure irq is low since acks messes with stuff
+        writePinLow(ITON_BT_IRQ_LINE);
+        #endif
+
+        switch (iton_bt_buffer[0]) {
+            case led_state:
+                iton_bt_led_state = iton_bt_buffer[1];
+                break;
+            case notification:
+                switch (iton_bt_buffer[1]) {
+                    case notif_battery:
+                        iton_bt_rx_battery_notif(iton_bt_buffer[2]);
+                        break;
+                    case notif_bluetooth:
+                        iton_bt_rx_bluetooth_notif(iton_bt_buffer[2]);
+                        break;
+                }
+                break;
+        }
+    }
+}
+#endif
+
+
+void iton_bt_data_cb(SPIDriver *spip) {
+    writePinLow(ITON_BT_IRQ_LINE);
+}
+
+/**
+ * Driver Functions
+ */
+void iton_bt_init(void) {
```

Comment:
a check needs to be done here. Some keyboards have a switch that not only toggles bt/usb but also kills power to the Iton module when the position is at USB. Check if the switch is defined and active before doing any kind of init. Triggering the switch should also call this to avoid hangs when switching to BT mode, since the device is effectively powering on cold. ( this should fix the need for a separate key needed to toggle between modes)

Reply:
that possibly related not to iton_bt_init only, but for the whole bluetooth init section